File: 142环形链表II.cpp
#include <iostream>
#include "list.h"
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        //if(!head||!head->next)return head;
        ListNode*p=head;
        ListNode*q=head;
        while(q&&q->next){
            p=p->next;
            q=q->next->next;
            if(p==q){
                ListNode*ans=head;
                while(ans!=p){
                ans=ans->next;
                p=p->next;
                }
                return ans;
            };
        }
        /*if(!p||!q)return nullptr;
        ListNode*ans=head;
        while(ans&&p){
            ans=ans->next;
            p=p->next;
            if(ans==p)break;
        }
        return ans;*/
        return nullptr;
    }
};

File: twopointer/142环形链表II.py
from typing import Optional
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        fast = slow = head
        while fast is not None and fast.next is not None:
            fast = fast.next.next
            slow = slow.next
            if fast == slow:
                p = fast
                q = head
                while p != q:
                    p = p.next
                    q = q.next
                return p
        return None